Improved Explodes trait and how artillery explodes

reaperrr opened the pull request:
Added `LoadedChance` to `Explodes`, changed how RA/TD artillery explodes.

See individual commits for more detailed descriptions.

reaperrr committed:
Clean up RA vehicles.yaml since UnitExplodeSmall is inherited from defaults

reaperrr committed:
Add descriptions and LoadedChance to Explodes trait

'Chance' only allows to control whether the actor explodes at all.
'LoadedChance' allows to control how likely the actor will explode violently using Weapon, and otherwise falls back to EmptyWeapon.

reaperrr committed:
Tweaked RA and TD artillery explosion

Renamed RA UnitExplode to AmmoExplode, added new UnitExplode with 1/10
damage.
Reduced TD UnitExplode damage to 1/10 of old value.

Now artillery always explodes, but there's only a 75% chance it explodes
violently and only when it has ammo.
If either condition isn't met, it explodes with the (now less violent) UnitExplode.

reaperrr commented:
I don't want to make that call on my own, but I think this should go into next playtest/release.

pchote commented:
> Reduced TD UnitExplode damage to 1/10 of old value.

This seems a bit drastic.  What's the balance motivation for such a large change?

reaperrr commented:
> This seems a bit drastic. What's the balance motivation for such a large change?

That this explosion currently isn't used at all in TD since #8303 removed it from artillery precisely because it dealt way too much (splash) damage.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -14,14 +14,22 @@
 		public readonly string Weapon = "UnitExplode";
-		[WeaponReference]
+
+		[WeaponReference, Desc("Weapon to use for explosion if no ammo/payload.")]
```
"`if no ammo/payload` ... is loaded."?

abcdefg30 commented:
Some thing about `LoadedChance` seems wrong to me...
If you set it to 0, the Weapon explosion will always occur.

Unit158 commented:
As it should. Loaded chance only changes the likeliness of a different weapon being used.

abcdefg30 commented:
`[Desc("Chance that the explosion will use Weapon if the actor has ammo/payload.")]`
Sounds to me that `Weapon` shouldn't be used if the chance is 0.

Unit158 commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -60,7 +68,8 @@ public void Killed(Actor self, AttackInfo e)
 		{
 			var shouldExplode = self.TraitsImplementing<IExplodeModifier>().All(a => a.ShouldExplode(self));
-			return shouldExplode ? info.Weapon : info.EmptyWeapon;
+			var useFullExplosion = self.World.SharedRandom.Next(100) > info.LoadedChance;
```
> should be <= :)

reaperrr commented:
Updated.

Mailaender commented on OpenRA.Mods.Common/Traits/Explodes.cs:
```diff
@@ -14,14 +14,22 @@
 		public readonly string EmptyWeapon = "UnitExplode";
 
+		[Desc("Chance that the explosion will use Weapon if the actor has ammo/payload.")]
+		public readonly int LoadedChance = 100;
```
This changes the previous default.

Mailaender replied:
Or wait probably not. This is just confusingly complicated.

Mailaender commented:
Stuff still explodes properly. :white_check_mark:

obrakmann commented:
The artillery in RA now makes the 'artillery firing sound' when it explodes. TD is fine.

reaperrr commented:
Updated.

- Reverted RA changes to `UnitExplode`, since it's the default explosion used by some traits (crashing aircraft, for example). This was not necessary for TD since it uses custom `HeliCrash` and `HeliExplode` for helicopters.
- Introduced an `ArtilleryExplosion` to RA instead, which is now used as `Weapon` when artillery explodes. 
- Artillery `EmptyWeapon` is now `UnitExplodeSmall` like for all other vehicles.

obrakmann commented:
Looks ok now. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/82f45c2533da91fd4efb8e4f3de023331c2ec8c4%5E...82f45c2533da91fd4efb8e4f3de023331c2ec8c4)
